### PR TITLE
Add internationalisation UI to modeladmin

### LINKED ISSUE #7856: Add internationalisation UI to modeladmin
#6230 set to add the locale filter on the modeladmin index, but was closed and it looks the changes never made it in

Needs:

- [ ] Locale switcher/filter on modeladmin index (most of the code in #6230)
- [ ] Locale switcher on modeladmin edit

### PR DESCRIPTION
**Pull request by andresmrm**
Initial work to fix #7856.

Merely copied modeladmin code from #6230, but it didn't work.

So I added a second commit that made the select show up in the modeladmin index, but using it still doesn't change the listed items... =(

### CODE REVIEW

File: wagtail/contrib/modeladmin/views.py [hunk 1]
```diff
@@ -254,7 +256,8 @@ class IndexView(SpreadsheetExportMixin, WMABaseView):
     SEARCH_VAR = 'q'
     ERROR_FLAG = 'e'
     EXPORT_VAR = 'export'
-    IGNORED_PARAMS = (ORDER_VAR, ORDER_TYPE_VAR, SEARCH_VAR, EXPORT_VAR)
+    LANG_VAR = 'lang'
```
Comment (zerolab) [inline wagtail/contrib/modeladmin/views.py:259]: you need to filter the queryset based on `LANG_VAR` (after https://github.com/wagtail/wagtail/blob/672482ee09663e62df138901c375d9988d72276e/wagtail/contrib/modeladmin/views.py#L297) to actually filter things.

I would also change from `lang` to `locale` to be consistent with the other places (e.g. snippets)

Comment (andresmrm) [reply wagtail/contrib/modeladmin/views.py:259]: Hmmm, is that the use of that `LANG_VAR`? I just copied it without understanding it. hehe

Now that you said that, I don't think we need the select or to filter the queryset here, because this PR already adds this (copied from #6230):
```        
if issubclass(self.model, TranslatableMixin):
    list_filter += ('locale', )
```
Which gives us this:

image

Unlike the snippets, the modeladmin already has this type of filter, so it seems better to use it.

So what is still missing now is a way to set the locale when adding a new item. Without it, it's always using the same locale, without asking. But I'm not sure how you would like it implemented. Maybe a select in the top bar like in the snippets?

Comment (zerolab) [reply wagtail/contrib/modeladmin/views.py:259]: yes, create/edit should follow the pattern used by snippets (selector in the top bar, handle the `locale` get param)

**Comment by andresmrm** [comment]
Added the same select used by the snippets to the modeladmin create, and used the ?locale= value when saving the instance. Not sure if I did it the right way...

Didn't changed the edit page, so for now you can only set the locale when creating, but can't edit it later.

The UI flow is a bit awkward, because every time you open the create page the locale will be set to the default one, and if you are adding multiple non default locale items you need to select the right locale every time. Maybe it should maintain the ?locale= value when going back to the index after adding an item. Not sure how to do it...

**Comment by andresmrm** [comment]
As I said, the functionality is not perfect, but at least it would make it minimally usable. It would be great if you could add it to the next version. :smiley: 
I don't know how robust is the test coverage for this project, but it would be good to make sure this PR isn't breaking anything else, I only marginally tested it. :grin: 

Also, if you think it would be mandatory to address the remaining problems I described for it to make to the next version, please confirm if what I did is in the right path and I can try to continue it today.

I understand you must be in a hurry to pack the next version, so I understand if this PR doesn't make it. :wink:

**Comment by zerolab** [comment]
Hey @andresmrm thank you for restarting work on this.

I've made further tweaks in https://gist.githubusercontent.com/zerolab/d50698800aaa25fb9e72011cdccde8c4/raw/3ad13f9b9394d67769dc15003802686855fd08ed/7857.patch -- tidies up the switcher on the create view, adds it to the edit and inspect view, as well as adding a locale column in the index view

Did not have time to write up tests, but they would go in `contrib/modeladmin/tests/test_simple_modeladmin.py` -- there are tests for the various views. I would add a new test that generates two new languages (e.g. French and Portuguese) and the creates some entries for a chosen model (either `Author` or `Book`) then tests the functionality. e.g. 
- index view: with or without locale filter applied
- create view: has locale switcher, then chosen language matches that of the `locale` GET param
- update view: doesn't show the switcher for an instance without translations. only shows the locale for the translation
- index view, same as update view

**Comment by andresmrm** [comment]
Hello! Merged your patch. I'll try to do the tests.

**Comment by andresmrm** [comment]
Changed your patch to avoid error. Not sure if it should be `if issubclass(self.model, TranslatableMixin) and self.locale:` instead of just `if issubclass(self.model, TranslatableMixin):` as I did...

**Comment by andresmrm** [comment]
For the tests, we need a model with `TranslatableMixin`, right? If I add it to the `Book` model in `wagtail/tests/modeladmintest/models.py` it would could affect the non i18n tests... Should I add a new model to `wagtail/tests/modeladmintest/models.py`?

**Comment by zerolab** [comment]
@andresmrm ah, you're right. probably a new simple model makes more sense.

on the question of `self.locale` - the property is on the view, not the model (accessible via `self.model`). Looks like it did not come through in the patch, but I have `self.locale = None` in `WMABaseView.__init__()` which does away with the error

File: wagtail/admin/templates/wagtailadmin/admin_base.html [hunk 2]
```diff
@@ -33,6 +33,12 @@
             {% locales as locales %}
             wagtailConfig.I18N_ENABLED = {% if i18n_enabled %}true{% else %}false{% endif %};
             wagtailConfig.LOCALES = {{ locales|safe }};
+            
+            // The locale of the current page/snippet being edited
+            wagtailConfig.ACTIVE_CONTENT_LOCALE = null;
```
Comment (zerolab) [inline wagtail/admin/templates/wagtailadmin/admin_base.html:38]: this should not be needed since we're not making use of any choosers

**Comment by andresmrm** [comment]
Ops, sorry, my confusion. Deleted my last commit and added `self.locale = None`.

**Comment by andresmrm** [comment]
Added tests for index and create.

> * update view: doesn't show the switcher for an instance without translations. only shows the locale for the translation
> * index view, same as update view

Not sure about what do you mean with "instance without translations", or "only ... the translation". Is the `TranslatableMixin.translation_key` used to link multiple translations of the same "thing"? When editing a post a see the "translate" button, but it isn't present when editing a modeladmin...

**Comment by zerolab** [comment]
Say you create a Book in English and leave it at that. `Book.get_translations()` (https://github.com/wagtail/wagtail/blob/093d59d4e3f8057f4bdb281f9d8cf223c90cf0c5/wagtail/core/models/i18n.py#L133) will return an empty queryset. So when you go to edit it, there should not be any switcher, just the current language

Having a "Translate" button when there are no translations would be pretty neat, but it expands the scope of this PR. Plus, if you edit a page without any translations, there is no button to start the translation. So I would say let's keep that for a follow up PR

**Comment by andresmrm** [comment]
Not sure if I understand it...
At the moment the edit page doesn't show the top right locale switcher for me. That seems right, since my items aren't linked to other items, as translations.
But without the "translate" button, it seems there is no way to link items as translations, so the switcher shouldn't appear anyway (unless someone adds then programmatically...).
Should we leave it this way? Or should I try to add the switcher when there are translations?

File: wagtail/contrib/modeladmin/views.py [hunk 3]
```diff
@@ -704,6 +739,20 @@ def dispatch(self, request, *args, **kwargs):
             return redirect(
                 self.url_helper.get_action_url('edit', self.pk_quoted)
             )
+
+        if self.locale:
```
Comment (zerolab) [inline wagtail/contrib/modeladmin/views.py:743]: should be 

```suggestion
        if getattr(settings, 'WAGTAIL_I18N_ENABLED', False) and issubclass(self.model, TranslatableMixin):
```

since this is the pattern used everywhere else

File: wagtail/contrib/modeladmin/views.py [hunk 4]
```diff
@@ -68,6 +70,7 @@ def __init__(self, model_admin):
         self.is_pagemodel = model_admin.is_pagemodel
         self.permission_helper = model_admin.permission_helper
         self.url_helper = model_admin.url_helper
+        self.locale = None
```
Comment (zerolab) [inline wagtail/contrib/modeladmin/views.py:73]: not needed since you're using the `if getattr(settings, 'WAGTAIL_I18N_ENABLED', False) and issubclass(self.model, TranslatableMixin):` everywhere

**Review by zerolab — CHANGES_REQUESTED** [review]

**Comment by zerolab** [comment]
https://github.com/wagtail/wagtail/pull/7857/files#diff-f369e817eb6a93bb0b23639713e726aec704aba1b4342d1aae8afeb625530ed9R1 and https://github.com/wagtail/wagtail/pull/7857#discussion_r788959001 will ensure the locale switcher shows in the edit view if you have any translations for the instance you are currently editing.

so the test should be something like:

```
book = TranslatableBook.objects.first()  # the english book

# get edit view, check no switcher shows

book.copy_for_translation(locale=Locale.objects.get(language_code='fr'))

# get edit view, check locale switcher shows
```

File: wagtail/contrib/modeladmin/views.py (hunk 4 above)
Comment (andresmrm) [reply wagtail/contrib/modeladmin/views.py]: Ok!

**Comment by andresmrm** [comment]
Hi, again!

Added the test. Had to add https://github.com/wagtail/wagtail/pull/7857/files#diff-131da701ebff77c03de56bb7ff4b43d307531db394c9905efbc57ca418ce6a59R177 without it I was getting `TypeError: Direct assignment to the reverse side of a related set is prohibited. Use +.set() instead.`. Not sure if this is the right way to fix it.

File: wagtail/core/models/i18n.py [hunk 5]
```diff
@@ -174,7 +174,7 @@ def copy_for_translation(self, locale):
 
         Note that the copy is initially unsaved.
         """
-        translated, child_object_map = _copy(self)
+        translated, child_object_map = _copy(self, exclude_fields=['index_entries'])
```
Comment (zerolab) [inline wagtail/core/models/i18n.py:177]: This rings a bell, but am curious as to why the change was needed

Comment (zerolab) [reply wagtail/core/models/i18n.py:177]: ah, are you hitting https://github.com/wagtail/wagtail-localize/issues/311 ?

Comment (andresmrm) [reply wagtail/core/models/i18n.py:177]: Yes! It seems the same problem.

**Comment by andresmrm** [comment]
Added `postgres_index_entries` to the excluded fields, looks like it worked. :smiley:

File: wagtail/core/models/i18n.py (hunk 5 above)
Comment (zerolab) [reply wagtail/core/models/i18n.py]: spoke with @kaedroho and we should filter out all generic relations (which would include `index_entries` and `postgres_index_entries`)
will try to add further notes on this, but it may not be until Wednesday due to some travel

Comment (andresmrm) [reply wagtail/core/models/i18n.py]: I already added `postgres_index_entries` to fix the postgres tests.
There is still a failing test, but it seems unrelated to this PR...

Comment (zerolab) [reply wagtail/core/models/i18n.py]: the failing test is against Django's `main` branch which is an allowed failure. Unfortunately GitHub Actions doesn't have a nice way to mark expected failures..

Comment (zerolab) [reply wagtail/core/models/i18n.py]: related issue for `excluded_fields` - #7735

Comment (andresmrm) [reply wagtail/core/models/i18n.py]: Good! Do I need to change anything else? Or just wait?

Comment (zerolab) [reply wagtail/core/models/i18n.py]: Hey @andresmrm there should not be anything else to change.
Will re-review tomorrow. Thank you

Comment (jsma) [reply wagtail/core/models/i18n.py]: This may be addressed by #8049, which will automatically exclude ``GenericRelation`` during copy operations.

**Comment by andresmrm** [comment]
Hi! Any update on this? It has been more than a month...

**Comment by zerolab** [comment]
Hey @andresmrm,

First off, sorry for the delay. It did not make the cut for 2.16, but it is on my list to review. 

Would you be able to rebase your branch on latest main, and lint it as well as we added black linting. https://github.com/wagtail/wagtail/wiki/Updating-PRs-for-linting-changes has some instructions

And I commit to reviewing the PR this weekend.

**Comment by zerolab** [comment]
@andresmrm tomorrow I have dedicated Wagtail core time and this is first on my list (I did promise I will review after the rebase, so that'll happen)

**Comment by andresmrm** [comment]
Hi! Any updates? Sooner or later I'll need this feature in a website I'm developing. :grin:

**Comment by zerolab** [comment]
Hey @andresmrm,

well, the PR is a pinned tab in my browser. Aside from that, I've been ill and out of action, then playing catch up.
On the other hand this will make it the next release in a bit over a month -- and we'll get that in,

I am sorry I promised and not got to it, especially this being your first contribution to Wagtail.

**Comment by andresmrm** [comment]
Thanks for the quick reply!
I'm sorry to know you've been will... I wish you the best recover possible!
But I'm glad to know this feature will make it to the next release, that should be good enough for us. :smiley:

**Comment by zerolab** [comment]
Thank you very much for this @andresmrm!
Merged in 483b7d2 and scheduled for release in 3.0

**Comment by andresmrm** [comment]
Great news! Thanks! :smile:
